In [2]:
import unicodecsv

**Loading data using unicodecsv library.**

In [5]:
data = []
my_file = open("data.csv", "rb")
reader = unicodecsv.DictReader(my_file)

for row in reader:
    data.append(row)
    
my_file.close()
    
    
for i in range(0, 5):
    print(data[i])

OrderedDict([('learner_id', '39'), ('country', 'PL'), ('in_course', 't'), ('unit', '1'), ('avg_score', '0.2'), ('completion', '0.1'), ('inv_rate', '0')])
OrderedDict([('learner_id', '39'), ('country', 'PL'), ('in_course', 't'), ('unit', '3'), ('avg_score', '1'), ('completion', '0.087'), ('inv_rate', '0')])
OrderedDict([('learner_id', '39'), ('country', 'PL'), ('in_course', 't'), ('unit', 'REVIEW 2'), ('avg_score', '0.1'), ('completion', '0.273'), ('inv_rate', '0')])
OrderedDict([('learner_id', '41'), ('country', 'PL'), ('in_course', 't'), ('unit', '1'), ('avg_score', '0.877'), ('completion', '0.35'), ('inv_rate', '0.381')])
OrderedDict([('learner_id', '41'), ('country', 'PL'), ('in_course', 't'), ('unit', '3'), ('avg_score', '0'), ('completion', '0.087'), ('inv_rate', '0')])


I think it may be useful to know the total number of students (student ids to be more precise).

In [11]:
id_set = set()
for data_point in data:
    id_set.add(data_point['learner_id'])
    
print (len(id_set))
print (len(data))

13158
81432


It will be also good to get the information about set of countries that students are from (and how many students from each country), how many courses are being taken with teacher or not, how many different units are there, so lets look into it:

In [15]:
country_set = set()
tCounter = 0
unit_set = set()
for data_point in data:
    country_set.add(data_point['country'])
    unit_set.add(data_point['unit'])
    if data_point['in_course'] == 't':
        tCounter+=1
    
print (country_set, len(country_set))
print (tCounter)
print (unit_set, len(unit_set))



{'', 'NZ', 'EC', 'CO', 'SV', 'BG', 'AU', 'AN', 'BY', 'CD', 'AF', 'KW', 'YE', 'LT', 'BD', 'AI', 'CR', 'TM', 'LV', 'AL', 'NL', 'ID', 'AZ', 'ZW', 'TC', 'GB', 'AT', 'IQ', 'CK', 'IT', 'BR', 'JP', 'SZ', 'DE', 'RO', 'AS', 'HK', 'MA', 'PL', 'CZ', 'PH', 'TL', 'ES', 'BL', 'PT', 'AO', 'VN', 'DZ', 'LY', 'ML', 'OM', 'FR', 'AG', 'CL', 'CY', 'RU', 'PM', 'MX', 'VA', 'SO', 'AR', 'SK', 'IR', 'MK', 'AX', 'GR', 'BE', 'CA', 'US', 'HR', 'KR', 'MD', 'AQ', 'CH', 'CN', 'GM', 'QU', 'PS', 'HU', 'UA', 'TN', 'VE', 'AD', 'TR', 'TH', 'TW', 'SA'} 87
75332
{'1', '7', '5', 'REVIEW 3', '10', '6', 'REVIEW 4', '9', 'VIDEO PODCASTS', '8', '12', '4', '11', 'REVIEW 2', '2', 'REVIEW 1', '3'} 17


I've just learned, that there are some students, that don't have anything in "country" field, so it would be good thing to investigate how many of them, and if is there something odd with these students. It can also be useful to build a dictionary with student: country and see if it can be a way to fix missing pieces of data (maybe some students have country in different entry?). It will be also a good way to find out if very student has only one country (probably should).

Also it will be great to know how many students are from each country, since it may be an important variable in the future.

There are 81432 entries in data, and majority of them has (75332) has "in_course" = "t". so there is small number of entries that are without the teacher, but stil a significant one - another thing worth investigating.

Since we know all users id by now, maybe it is also a good moment to see how many and what units were completed by each students - so lets create another dictionary with student : [units].


In [37]:
mCount = 0;
id_and_country = {}

country_numbers = {}
no_teacher = []

student_and_unit = {}

for country in country_set:
    country_numbers[country] = 0
    
for student in id_set:
    student_and_unit[student] = []

for data_point in data:
    if data_point['learner_id'] not in id_and_country:
        id_and_country[data_point['learner_id']] = data_point['country']
    if data_point['learner_id'] in id_and_country:
        if id_and_country[data_point['learner_id']] != data_point['country']:
            print("Student has two different countries", datapoint['learner_id']) 
    country_numbers[data_point["country"]]+=1
    if data_point['country'] == '':
        print(data_point)
        mCount +=1
    if data_point['in_course'] != 't':
        no_teacher.append(data_point)

print (mCount)

#print (id_and_country['802'])
#print (id_and_country['5811'])
    
print (country_numbers)

print (country_numbers['TR']/len(data))

print ((country_numbers['TR'] + country_numbers['PL'] + country_numbers['CO'] + country_numbers['ES'] 
       + country_numbers['IT'])/len(data))
    
print(no_teacher[0:10])        


#    try to find how many units did every student and which ones, maybe sort these?

OrderedDict([('learner_id', '802'), ('country', ''), ('in_course', 't'), ('unit', '1'), ('avg_score', '0.954'), ('completion', '1'), ('inv_rate', '0')])
OrderedDict([('learner_id', '802'), ('country', ''), ('in_course', 't'), ('unit', '2'), ('avg_score', '0.877'), ('completion', '1'), ('inv_rate', '0')])
OrderedDict([('learner_id', '802'), ('country', ''), ('in_course', 't'), ('unit', '3'), ('avg_score', '0.883'), ('completion', '0.739'), ('inv_rate', '0')])
OrderedDict([('learner_id', '802'), ('country', ''), ('in_course', 't'), ('unit', 'REVIEW 1'), ('avg_score', '0.904'), ('completion', '1'), ('inv_rate', '0.141')])
OrderedDict([('learner_id', '802'), ('country', ''), ('in_course', 't'), ('unit', 'VIDEO PODCASTS'), ('avg_score', '0.955'), ('completion', '0.175'), ('inv_rate', '0')])
OrderedDict([('learner_id', '5811'), ('country', ''), ('in_course', 't'), ('unit', '1'), ('avg_score', '0.967'), ('completion', '0.3'), ('inv_rate', '0')])
OrderedDict([('learner_id', '5811'), ('country'

So there are only 8 entries that are missing country, and all of them belong to only two different students. Also, there are no students, that have more than one country. I've also checked if I can fix missing countries from these two students. Unfortunately both tries return a "key error" which means there is no chance to reconstruct countries of these two students.

It looks like the most of the data comes from Turkey (52727), and it is ver close to 65%. There are also 5925 entries from Poland, 8409 from Spain, and 5537 from Colombia. The only other country with more than 1000 entries is Italy (1779). Entries from these 5 countries are more than 91% of data.

I've also looked at some data from students that are not  guided by teachers ("in_course" = 'f'), but it could be also good to see if they are only doing units like this, or they sometimes  have "t" and sometimes "f" here.

// next: some plots with entries / countries, student / countries, average units from student from each ocuntry, average entries from teacher and no teacher

// teacher vs no_teacher inv_rate, top 5 countries
